File: 2022/day6.cpp
#include <bit>
#include <cinttypes>
#include <chrono>
#include <fstream>
#include <iostream>
#include <vector>

const size_t BATCH_SIZE = 64;
const size_t BATCH_NUM = 4096/BATCH_SIZE;

bool substring_unique_copyless(std::vector<uint8_t> *s, size_t right, size_t length) {
	for (size_t i = (right-length); i<right; i++) {
		uint8_t c = (*s)[i];
		for (size_t j = (i+1); j<right; j++) {
			if ((*s)[j] == c) {
				return false;
			}
		}
	}
	return true;
}

size_t find_first_unique_run(std::vector<uint8_t> *s, size_t num, size_t skip) {
	for (size_t right = skip; right<s->size(); right++) {
		if (substring_unique_copyless(s, right, num)) {
			return right;
		}
	}
	return 0;
}


std::pair<size_t, size_t> find_first_unique_runs(std::vector<uint8_t> *s) {
	size_t four = 0;
	size_t fourteen = 0;
	uint32_t masks[4096];
	uint32_t scratch_masks[BATCH_SIZE];
	uint8_t scratch_masks_bits[BATCH_SIZE];
	for (int i=0; i<4096; i++) {
		masks[i] = 1 << ((*s)[i]-'a');
	}
	// Turn masks into masks2. Indices are now +1.
	for (int i=0; i<4095; i++) {
		masks[i] |= masks[i+1];
	}
	for (int batch=1; batch<BATCH_NUM; batch++) {
		for (int i=0; i<BATCH_SIZE; i++) {
			int idx = batch*BATCH_SIZE + i;
			scratch_masks[i] = masks[idx] | masks[idx-2];
			scratch_masks_bits[i] = std::popcount(scratch_masks[i]);
		}
		for (int i=0; i<BATCH_SIZE; i++) {
			if (scratch_masks_bits[i] == 4) {
				four = batch*BATCH_SIZE + i;
				goto four_found;
			}
		}
	}
four_found:
	// Turn masks2 into masks4. Indices are now +3.
	for (int i=four; i<4093; i++) {
		masks[i] |= masks[i+2];
	}
	// Turn masks4 into masks8. Indices are now +7.
	for (int i=fourteen; i<4089; i++) {
		masks[i] |= masks[i+4];
	}
	for (int batch=(four/BATCH_SIZE); batch<BATCH_NUM; batch++) {
		for (int i=0; i<BATCH_SIZE; i++) {
			int idx = batch*BATCH_SIZE + i;
			scratch_masks[i] = masks[idx] | masks[idx-6];
			scratch_masks_bits[i] = std::popcount(scratch_masks[i]);
		}
		for (int i=0; i<BATCH_SIZE; i++) {
			if (scratch_masks_bits[i] == 14) {
				fourteen = batch*BATCH_SIZE + i;
				goto fourteen_found;
			}
		}
	}
fourteen_found:
	return std::pair<size_t, size_t>(four+2, fourteen+8);
}

int main(int argc, char **argv) {
	size_t iterations = 1000000;
	auto filename = "input/6";
	switch(argc) {
		case 3:  // a.out filename iterations
			iterations = std::stoi(argv[2]);
			// std::cout << "Set iterations to: " << iterations << '\n';  // fallthrough
		case 2:  // a.out filename
			filename = argv[1];
			// std::cout << "Set filename to: " << filename << '\n';
		default:
			break;
	}

	auto input_file = std::ifstream(filename, std::ios::binary | std::ios::ate);
	const std::streamoff eof_position = static_cast<std::streamoff>(input_file.tellg());
	auto s = std::vector<uint8_t>(eof_position);
	input_file.seekg(0, std::ios::beg);
	input_file.read(reinterpret_cast<char*>(s.data()), eof_position);

	size_t volatile four = 0;  // Don't you dare optimize out the iterations!
	size_t volatile fourteen = 0;
	std::cout << "Running " << iterations << " iterations with batch size " << BATCH_SIZE << ":\n";
	auto clock = std::chrono::high_resolution_clock();
	auto t0 = clock.now();
	for (size_t i = 0; i<iterations; i++) {
		// four = find_first_unique_run(&s, 4, 4);
		// fourteen = find_first_unique_run(&s, 14, four);
		auto result = find_first_unique_runs(&s);
		four = result.first;
		fourteen = result.second;
	}
	auto t1 = clock.now();
	auto delta = t1 - t0;
	auto per_iteration = delta / iterations;
	std::cout << "Part 1: " << four << '\n';
	std::cout << "Part 2: " << fourteen << '\n';
	std::cout << "Completed " << iterations << " iterations in "
		<< std::chrono::duration_cast<std::chrono::milliseconds>(delta).count()
		<< "ms (~"
		<< std::chrono::duration_cast<std::chrono::nanoseconds>(per_iteration).count()
		<< "ns each)\n";
}

File: 2022/day6.rs
#![allow(dead_code)]
use std::env;
use std::fs;
use std::time::Instant;

type Int = usize;

fn substring_unique_copyless(s: &Vec<u8>, right: Int, length: Int) -> bool {
	for i in (right-length)..right {
		let c = s[i];
		for j in (i+1)..right {
			if s[j] == c {
				return false;
			}
		}
	}
	return true;
}

fn find_first_unique_run(s: &Vec<u8>, num: Int, skip: Int) -> Int {
	for right in num.max(skip)..s.len() {
		if substring_unique_copyless(s, right, num) {
			return right;
		}
	}
	return 0;
}

fn mask_to_string(mask: u32) -> String {
	let mut v: Vec<u8> = Vec::with_capacity(mask.count_ones() as usize);
	for i in 0..26 {
		if mask & (1<<i) != 0 {
			v.push(b'a'+i);
		}
	}
	return String::from_utf8(v).unwrap();
}

fn find_first_unique_runs(s: &Vec<u8>) -> (Int, Int) {
	const BATCH_SIZE: Int = 8;
	const BATCH_NUM: Int = 4096/BATCH_SIZE;
	let mut four: Int = 0;
	let mut fourteen: Int = 0;
	let mut masks: [u32; 4096] = [0; 4096];
	let mut scratch_masks: [u32; BATCH_SIZE] = [0; BATCH_SIZE];
	let mut scratch_masks_bits: [u8; BATCH_SIZE] = [0; BATCH_SIZE];
	for i in 0..4096 {
		masks[i] = 1 << (s[i]-b'a');
	}
	// Turn masks into masks2. Indices are now +1.
	for i in 0..(4096-1) {
		masks[i] |= masks[i+1];
	}
	'four_loop: for batch in 1..(BATCH_NUM) {
		for i in 0..BATCH_SIZE {
			let idx = batch*BATCH_SIZE + i;
			scratch_masks[i] = masks[idx] | masks[idx-2];
			scratch_masks_bits[i] = scratch_masks[i].count_ones() as u8;
		}
		for i in 0..BATCH_SIZE {
			if scratch_masks_bits[i] == 4 {
				four = batch*BATCH_SIZE + i;
				break 'four_loop;
			}
		}
	}
	// Turn masks2 into masks4. Indices are now +3.
	for i in four..(4096-3) {
		masks[i] |= masks[i+2];
	}
	// Turn masks4 into masks8. Indices are now +7.
	for i in four..(4096-7) {
		masks[i] |= masks[i+4];
	}
	'fourteen_loop: for batch in (four/BATCH_SIZE)..(BATCH_NUM) {
		for i in 0..BATCH_SIZE {
			let idx = batch*BATCH_SIZE + i;
			// scratch_masks[i] = masks[idx] | masks[idx-4] | masks[idx-8] | masks[idx-10];
			scratch_masks[i] = masks[idx] | masks[idx-6];
			scratch_masks_bits[i] = scratch_masks[i].count_ones() as u8;
		}
		for i in 0..BATCH_SIZE {
			if scratch_masks_bits[i] == 14 {
				fourteen = batch*BATCH_SIZE + i;
				break 'fourteen_loop;
			}
		}
	}
	return (four+2, fourteen+8);
}

fn main() {
	let mut filename = "input/6";
	let mut iterations = 1_000_000;
	let args: Vec<String> = env::args().collect();
	match args.len() {
		3 => {filename = &args[1]; iterations = args[2].parse().expect(&("Invalid number for iterations: ".to_owned() + &args[2]))},
		2 => filename = &args[1],
		_ => (),
	};
	let s: Vec<u8> = fs::read(filename).expect("Input not found!");

	println!("Running {} iterations:", iterations);
	let t0 = Instant::now();
	let mut four = 0;
	let mut fourteen = 0;
	for _ in 0..iterations {
		// four = find_first_unique_run(&s, 4, 0);
		// fourteen = find_first_unique_run(&s, 14, four);
		(four, fourteen) = find_first_unique_runs(&s);
	}
	let duration = t0.elapsed();
	let per_iteration = duration / iterations;
	println!("Part 1: {}", four);
	println!("Part 2: {}", fourteen);
	println!("Completed {} iterations in {:.3}s (~{}ns each)", iterations, duration.as_secs_f32(), per_iteration.as_nanos());
}
